_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -173,6 +173,54 @@ You can find detailed explanations, runnable examples and tips to maintain binar
 
 Again, we recommend using MiMa to double-check that you have not broken binary compatibility after making changes.
 
+### Changing case class definition in a compatible manner
+
+Sometimes it is desirable to be able to change the definition of a case class (adding and/or removing fields) while still staying compatible with the users of the case class, i.e. not breaking the so called _binary compatibility_.
+
+To achieve that, follow this pattern:
+ * make the constructor private
+ * define a private `unapply` function in the companion object (note that by doing that the case class loses the ability to be used in an extractor pattern match)
+ * define `withXXX` methods on the case class that create a new instance with the respective field changed
+ * define custom `apply` factory method(s) in the companion object (these can use the private constructor)
+
+Example:
+
+{% tabs case_class_compat_1 %}
+{% tab 'Scala 3 Only' %}
+
+```scala
+case class Person private (name: String, age: Int):
+  def withName(name: String) = copy(name = name)
+  def withAge(age: Int) = copy(age = age)
+object Person:
+  def apply(name: String, age: Int) = new Person(name, age)
+  private def unapply(p: Person) = p
+```
+{% endtab %}
+{% endtabs %}
+
+Later in time, you can ammend the original case class definition. You
+ * add a new field `address`,
+ * add a custom `withAddress` method and
+ * add an `apply` factory method to the companion.
+
+{% tabs case_class_compat_2 %}
+{% tab 'Scala 3 Only' %}
+```scala
+case class Person private (name: String, age: Int, address: String = ""):
+  ...
+  def withAddress(address: String) = copy(address = address)
+object Person:
+  ...
+  def apply(name: String, age: Int, address: String) = new Person(name, age, address)
+```
```

Comment:
IIRC, that change will still trigger a MiMa error because the primary constructor, although private from the point of view of Scala, was not private at the bytecode level (this is an “implementation detail”). Therefore, you have several options here:
- add back the previous constructor signature as a secondary constructor:
  ~~~ scala
  case class Person private (name: String, age: Int, address: Option[String]):
    def this(name: String, age: Int): Person = this(name, age, None)
  ~~~
- add an exclusion filter to the MiMa settings (in such a case, the filter is not 100% safe because there could be existing _Java_ code calling the previous constructor that will break when run with the new class definition)

Reply:
So it turns out we have to do 3 things now:
 * add a new constructor, private to the companion object, with the original fields
 * add a custom `withAddress` method and
 * add an `apply` factory method to the companion.

This pattern doesn't seem as attractive as it did in the beginning... :confused:

Reply:
Or: Fix MiMa to not flag this pattern. That would be the correct thing to do, which might also help other code.

Reply:
The problem in this case is that Java code would see this as a perfectly normal constructor. No funny name, no synthetic or artifact flag. Just your regular, public constructor.

Reply:
"Private" primary constructors vs MiMa seems like a topic that needs further discussion.
I've kicked off https://contributors.scala-lang.org/t/private-primary-constructor-vs-mima/6050